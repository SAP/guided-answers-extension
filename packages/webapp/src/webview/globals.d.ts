interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
