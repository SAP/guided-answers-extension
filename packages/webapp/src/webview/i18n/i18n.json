{
    "GUIDED_ANSWER_HEADER": "Guide",
    "SEARCH_GUIDED_ANSWERS": "Search Guided Answers",
    "ALL_ANSWERS": "Home",
    "STEP_BACK": "Step back",
    "RESTART": "Restart",
    "GUIDED_ANSWERS": "Guided Answers",
    "NO_ANSWERS_FOUND": "No answers found",
    "PLEASE_MODIFY_SEARCH": "Please modfiy search",
    "THIS_SOLVED_MY_ISSUE": "This solved my issue",
    "THIS_DID_NOT_SOLVED_MY_ISSUE": "This did not solve my issue",
    "THANK_YOU_TEXT": "We are glad to hear that your issue has been resolved and we hope you enjoy using the Guided Answers application.",
    "THANKS": "Thanks!",
    "HOME": "Home",
    "SEND": "Send",
    "CLOSE": "Close",
    "ISSUE_IS_NOT_RESOLVED": "Issue is not resolved",
    "WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED": "We are sorry to hear that your issue is not yet resolved.",
    "PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL": "Please tell us if this answer was helpful",
    "THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE": "There are several options for getting further assistance:",
    "SEARCH_FOR_ANOTHER_GUIDED_ANSWER": "Search for another Guided Answer",
    "START_AN_EXPERT_CHAT": "Start an Expert Chat",
    "SCHUEDULE_AN_EXPERT": "Schedule an Expert",
    "OPEN_AN_INCIDENT": "Open an Incident",
    "ASK_THE_SAP_COMMUNITY": "Ask the SAP Community",
    "FEEDBACK": "Feedback",
    "FEEDBACK_DIALOG_TITLE": "Is this content helpful?",
    "FEEDBACK_DIALOG_SUBTEXT": "If you have suggestions on how to improve this content we would love to hear them!",
    "FEEDBACK_DIALOG_DISCLAIMER": "Your feedback is anonymous, we do not collect any personal data.",
    "FEEDBACK_DIALOG_SUGGESTION": "Your suggestion",
    "FEEDBACK_SENT_DIALOG_TITLE": "Message sent",
    "FEEDBACK_SENT_DIALOG_SUBTEXT": "Thanks for the feedback!"
}
