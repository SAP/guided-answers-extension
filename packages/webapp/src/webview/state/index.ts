export { actions, store } from './store';
export { getInitialState } from './reducers';
