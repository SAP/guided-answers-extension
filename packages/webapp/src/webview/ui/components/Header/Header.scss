.guided-answer {
    &__header {
        display: flex;
        width: 100%;
        z-index: 10;
        background-color: var(--vscode-editor-background);
        position: sticky;
        top: 0;
        padding-bottom: 20px;
        &__sub {
            display: flex;
            align-items: center;
            width: 100%;
        }
        &__title {
            font-weight: 600;
            margin: 0;
            color: var(--vscode-settings-headerForeground);
            line-height: 29px;
        }
        &__subtitle {
            color: var(--vscode-editorCodeLens-foreground, #999);
            font-size: 2em;
        }
        &__navButtons {
            height: 25px;
            margin-top: 30px;
            color: var(--vscode-foreground);
            padding: 0 5px 0px 5px;
            text-align: start;
            cursor: pointer;
            background: none;
            border: 0;
            &__content {
                vertical-align: middle;
                display: inline-block;
                &__text {
                    margin-left: 3px;
                }
            }
            display: flex;
            align-items: center;
            svg {
                width: 16px;
                height: 16px;
            }
        }
        &__navButtons:hover {
            background: var(--vscode-menubar-selectionBackground);
            outline: 1px solid var(--vscode-contrastActiveBorder);
            border-radius: 3px;
            > span {
                text-decoration: none;
            }
        }
        &__navButtons:focus-visible {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
        }

        &__allAnswersButton {
            width: 25%;
            margin-right: 80px;
        }
        &__back-restart-buttons {
            display: flex;
            justify-content: space-between;
            width: 75%;
            max-width: 620px;
        }
        &__restart-feedback-buttons {
            display: flex;
        }
        &__divider {
            border-left: 1px solid var(--vscode-editorWidget-border);
            height: 16px;
            margin-top: 35px;
            margin-left: 10px;
            margin-right: 10px;
        }
        &__logoAndTitle {
            display: flex;
            align-items: center;
            margin-top: 30px;
            min-width: 280px;
        }
        &__searchField {
            display: flex;
            align-items: start;
            flex-direction: row;
            margin-top: 30px;
            width: 66%;
        }
    }
}

.home-icon {
    width: 62px;
}

@media only screen and (max-width: 625px) {
    .guided-answer {
        &__header {
            &__logoAndTitle {
                width: 50%;
            }

            &__sub {
                align-items: start;
                flex-direction: column;
            }

            &__searchField {
                margin-top: 20px;
                align-items: start;
            }
            &__navButtons {
                text-align: center;
                justify-content: center;
            }
            &__navButtons__content {
                margin-left: 0;
            }
            &__navButtons__content__text {
                display: none;
            }
            &__back-restart-buttons {
                justify-content: space-between;
                width: 100%;
            }
            &__allAnswersButton {
                margin-right: 48px;
                width: 28px;
                margin-left: 6px;
            }
        }
    }
    .home-icon {
        width: 28px;
        padding: 0 3px 2px 3px;
        border-radius: 3px;
        color: var(--vscode-menu-foreground, #cccccc);
    }
    .home-icon:hover {
        background: var(--vscode-menubar-selectionBackground);
    }
}
