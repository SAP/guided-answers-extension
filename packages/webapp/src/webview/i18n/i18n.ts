import { use } from 'i18next';
import { initReactI18next } from 'react-i18next';
import * as i18nEn from './i18n.json';

export { i18n } from 'i18next';

export function initI18n(): void {
    use(initReactI18next)
        .init({
            resources: {
                en: {
                    translation: i18nEn
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            interpolation: {
                escapeValue: false
            }
        })
        .catch((error) => console.error(error));
}
