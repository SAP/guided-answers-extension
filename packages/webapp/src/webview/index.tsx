import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
