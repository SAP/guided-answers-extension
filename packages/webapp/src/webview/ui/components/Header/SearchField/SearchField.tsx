import React from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../../types';
import { actions } from '../../../../state';
import { VSCodeTextField } from '@vscode/webview-ui-toolkit/react';
import { Filters } from '../Filters';

let timer: NodeJS.Timeout;
/**
 *
 * @returns An input field
 */
export function SearchField() {
    const appState = useSelector<AppState, AppState>((state) => state);
    return (
        <div className="guided-answer__header__searchField">
            <VSCodeTextField
                className="tree-search-field"
                value={appState.query}
                readOnly={appState.loading}
                placeholder="Search Guided Answers"
                id="search-field"
                onInput={(e: any) => {
                    const newValue = e?.target?.value;
                    if (newValue !== undefined) {
                        clearTimeout(timer);
                        actions.setQueryValue(newValue);
                        timer = setTimeout(() => {
                            actions.searchTree({
                                query: newValue,
                                filters: {
                                    product: [],
                                    component: []
                                },
                                paging: {
                                    responseSize: appState.pageSize,
                                    offset: 0
                                }
                            });
                        }, 400);
                    }
                }}></VSCodeTextField>
            <Filters />
        </div>
    );
}
