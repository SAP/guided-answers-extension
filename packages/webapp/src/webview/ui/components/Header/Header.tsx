import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../types';
import { AllAnswersButton, BackButton, GeneralFeedbackButton, RestartButton } from './NavigationButtons';
import { Logo } from './Logo';
import './Header.scss';
import { SearchField } from './SearchField';
import { FocusZone } from '@fluentui/react-focus';

/**
 * Renders and returns the header section.
 *
 * @param props Props for Header component
 * @param props.showSub Show/Hide subsection of the Header
 * @param props.showNavButons Show/Hide navigation buttons
 * @param props.showLogo Show/Hide SAP logo
 * @param props.showSearch Show/Hide Search field
 * @returns - react element for header section
 */
export function Header(props: {
    showSub: boolean;
    showNavButons: boolean;
    showLogo: boolean;
    showSearch: boolean;
}): ReactElement {
    const appState = useSelector<AppState, AppState>((state) => state);
    return (
        <div className="guided-answer__header" style={{ paddingBottom: props.showNavButons === true ? '0' : '20px' }}>
            {props.showSub === true ? (
                <>
                    <div className="guided-answer__header__sub">
                        {!!props.showLogo && <Logo />}
                        {!!props.showSearch && <SearchField />}
                    </div>{' '}
                </>
            ) : (
                <></>
            )}

            {props.showNavButons === true ? (
                <>
                    <FocusZone isCircularNavigation={true} className="guided-answer__header">
                        <div className="guided-answer__header__allAnswersButton">
                            <AllAnswersButton />
                        </div>

                        <div className="guided-answer__header__back-restart-buttons">
                            <BackButton />
                            <div className="guided-answer__header__restart-feedback-buttons">
                                {appState.activeGuidedAnswerNode.length > 1 && <RestartButton />}
                                {appState.activeGuidedAnswerNode.length > 1 && (
                                    <div className="guided-answer__header__divider"></div>
                                )}
                                <GeneralFeedbackButton />
                            </div>
                        </div>
                    </FocusZone>
                </>
            ) : (
                <></>
            )}
        </div>
    );
}
