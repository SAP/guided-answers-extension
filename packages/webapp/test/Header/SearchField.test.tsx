import React from 'react';
import { shallow } from 'enzyme';
import { SearchField } from '../../src/webview/ui/components/Header/SearchField';
import { initI18n } from '../../src/webview/i18n';
import { actions } from '../../src/webview/state';
import i18next from 'i18next';

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

jest.mock('../../src/webview/state', () => {
    return {
        actions: {
            searchTree: jest.fn(),
            setQueryValue: jest.fn()
        }
    };
});

jest.mock('@vscode/webview-ui-toolkit/react', () => ({
    VSCodeTextField: () => (
        <>
            <div>SearchField</div>
        </>
    )
}));

jest.mock('react-redux', () => {
    const lib = jest.requireActual('react-redux');
    const state = { query: 'Fiori tools' };
    return {
        ...lib,
        useSelector: () => state
    };
});

describe('<SearchField />', () => {
    let wrapper: any;
    initI18n();
    beforeEach(() => {
        wrapper = shallow(<SearchField />);
    });

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a SearchField component', () => {
        expect(wrapper.find('.guided-answer__header__searchField').length).toBe(1);
        expect(wrapper.find('VSCodeTextField').length).toBe(1);
        expect(wrapper.find('VSCodeTextField').at(0).props().onInput).toBeDefined();
        expect(wrapper.find('VSCodeTextField').at(0).props().id).toBe('search-field');
        expect(wrapper.find('VSCodeTextField').at(0).props().placeholder).toBe(i18next.t('SEARCH_GUIDED_ANSWERS'));

        //Test input event
        wrapper
            .find('VSCodeTextField')
            .at(0)
            .simulate('input', { target: { value: 'Fiori Tools' } });
        expect(actions.setQueryValue).toBeCalled();
        expect(setTimeout).toHaveBeenCalledTimes(1);

        // Fast-forward until all timers have been executed
        jest.runAllTimers();

        expect(actions.searchTree).toHaveBeenCalledTimes(1);
    });
});
