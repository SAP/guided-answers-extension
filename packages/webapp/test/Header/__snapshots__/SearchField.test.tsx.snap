// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchField /> Should render a SearchField component, on home screen 1`] = `
<div>
  <div
    class="guided-answer__header__searchField"
    id="search-field-container"
  >
    <div
      class="ms-SearchBox tree-search-field root-109"
    >
      <div
        aria-hidden="true"
        class="ms-SearchBox-iconContainer iconContainer-110"
      >
        <i
          aria-hidden="true"
          class="ms-SearchBox-icon icon-114"
          data-icon-name="Search"
        >
          <svg
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.5024,9.9999 C7.0214,9.9999 5.0024,7.9809 5.0024,5.4999 C5.0024,3.0189 7.0214,0.9999 9.5024,0.9999 C11.9834,0.9999 14.0024,3.0189 14.0024,5.4999 C14.0024,7.9809 11.9834,9.9999 9.5024,9.9999 M9.5024,-0.0001 C6.4704,-0.0001 4.0024,2.4679 4.0024,5.4999 C4.0024,6.9469 4.5684,8.2609 5.4854,9.2439 L0.0004,15.3309 L0.7434,16.0009 L6.2314,9.9109 C7.1474,10.5919 8.2764,10.9999 9.5024,10.9999 C12.5354,10.9999 15.0024,8.5319 15.0024,5.4999 C15.0024,2.4679 12.5354,-0.0001 9.5024,-0.0001"
              fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </div>
      <input
        class="ms-SearchBox-field field-113"
        id="search-field"
        placeholder="Search Guided Answers"
        role="searchbox"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`<SearchField /> Should render a SearchField component, on search screen 1`] = `
<div>
  <div
    class="guided-answer__header__searchField"
    id="search-field-container"
  >
    <div
      class="ms-SearchBox tree-search-field root-109"
    >
      <div
        aria-hidden="true"
        class="ms-SearchBox-iconContainer iconContainer-110"
      >
        <i
          aria-hidden="true"
          class="ms-SearchBox-icon icon-114"
          data-icon-name="Search"
        >
          <svg
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.5024,9.9999 C7.0214,9.9999 5.0024,7.9809 5.0024,5.4999 C5.0024,3.0189 7.0214,0.9999 9.5024,0.9999 C11.9834,0.9999 14.0024,3.0189 14.0024,5.4999 C14.0024,7.9809 11.9834,9.9999 9.5024,9.9999 M9.5024,-0.0001 C6.4704,-0.0001 4.0024,2.4679 4.0024,5.4999 C4.0024,6.9469 4.5684,8.2609 5.4854,9.2439 L0.0004,15.3309 L0.7434,16.0009 L6.2314,9.9109 C7.1474,10.5919 8.2764,10.9999 9.5024,10.9999 C12.5354,10.9999 15.0024,8.5319 15.0024,5.4999 C15.0024,2.4679 12.5354,-0.0001 9.5024,-0.0001"
              fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              fill-rule="evenodd"
            />
          </svg>
        </i>
      </div>
      <input
        class="ms-SearchBox-field field-113"
        id="search-field"
        placeholder="Search Guided Answers"
        role="searchbox"
        value=""
      />
    </div>
    <div
      id="filters"
    >
      <button
        aria-disabled="true"
        class="ms-Button ms-Button--icon is-disabled filter-button filter-button-selected root-115"
        data-is-focusable="false"
        disabled=""
        id="filter-products"
        title="Filter Products"
        type="button"
      >
        <span
          class="ms-Button-flexContainer flexContainer-116"
          data-automationid="splitbuttonprimary"
        >
          <i
            aria-hidden="true"
            class="ms-Icon root-105 ms-Button-icon icon-118"
            data-icon-name="Table"
          >
            <svg
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1,16 C0.44771525,16 0,15.5522847 0,15 L0,1 C0,0.44771525 0.44771525,0 1,0 L15,0 C15.5522847,0 16,0.44771525 16,1 L16,15 C16,15.5522847 15.5522847,16 15,16 L1,16 Z M5,11 L1,11 L1,15 L5,15 L5,11 Z M10,11 L6,11 L6,15 L10,15 L10,11 Z M15,11 L11,11 L11,15 L15,15 L15,11 Z M5,6 L1,6 L1,10 L5,10 L5,6 Z M10,6 L6,6 L6,10 L10,10 L10,6 Z M15,6 L11,6 L11,10 L15,10 L15,6 Z M5,1 L1,1 L1,5 L5,5 L5,1 Z M6,1 L6,5 L10,5 L10,1 L6,1 Z M11,5 L15,5 L15,1 L11,1 L11,5 Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              />
            </svg>
          </i>
        </span>
      </button>
      <button
        aria-disabled="true"
        class="ms-Button ms-Button--icon is-disabled filter-button filter-button-selected root-115"
        data-is-focusable="false"
        disabled=""
        id="filter-components"
        title="Filter Components"
        type="button"
      >
        <span
          class="ms-Button-flexContainer flexContainer-116"
          data-automationid="splitbuttonprimary"
        >
          <i
            aria-hidden="true"
            class="ms-Icon root-105 ms-Button-icon icon-118"
            data-icon-name="IdTag"
          >
            <svg
              fill="none"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6 2C6 1.44772 6.44775 1 7 1H9C9.55225 1 10 1.44772 10 2V3.58578C10 3.85101 9.89465 4.10535 9.70715 4.29289L8.70715 5.29289C8.31653 5.68342 7.68347 5.68342 7.29285 5.29289L6.29285 4.29289C6.10535 4.10535 6 3.85101 6 3.58578V2ZM7 2.5C7 2.22385 7.22388 2 7.5 2H8.5C8.77612 2 9 2.22385 9 2.5C9 2.77615 8.77612 3 8.5 3H7.5C7.22388 3 7 2.77615 7 2.5Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
                fill-rule="evenodd"
              />
              <path
                d="M5.04333 4H1C0.447754 4 0 4.44772 0 5V14C0 14.5523 0.447754 15 1 15H15C15.5522 15 16 14.5523 16 14V5C16 4.44772 15.5522 4 15 4H10.9567C10.8771 4.37594 10.6901 4.72412 10.4142 5H15V14H1V5H5.58582C5.30994 4.72412 5.12292 4.37594 5.04333 4Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              />
              <path
                clip-rule="evenodd"
                d="M3 7C2.44775 7 2 7.44772 2 8V11C2 11.5523 2.44775 12 3 12H5C5.55225 12 6 11.5523 6 11V8C6 7.44772 5.55225 7 5 7H3ZM5 8H3V11H5V8Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
                fill-rule="evenodd"
              />
              <path
                d="M7 7.5C7 7.22385 7.22388 7 7.5 7H13.5C13.7761 7 14 7.22385 14 7.5C14 7.77615 13.7761 8 13.5 8H7.5C7.22388 8 7 7.77615 7 7.5Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              />
              <path
                d="M7 9.5C7 9.22385 7.22388 9 7.5 9H13.5C13.7761 9 14 9.22385 14 9.5C14 9.77615 13.7761 10 13.5 10H7.5C7.22388 10 7 9.77615 7 9.5Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              />
              <path
                d="M11 11.5C11 11.2239 11.2239 11 11.5 11H13.5C13.7761 11 14 11.2239 14 11.5C14 11.7761 13.7761 12 13.5 12H11.5C11.2239 12 11 11.7761 11 11.5Z"
                fill="var(--vscode-icon-foreground, var(--vscode-foreground))"
              />
            </svg>
          </i>
        </span>
      </button>
    </div>
  </div>
</div>
`;
